### PR TITLE
NEW BACKEND! MongoDB Atlas

### PR DESCRIPTION
**Pull request by caseyclements**
## Description
This pull request introduces MongoDB integration as a document index backend, enhancing search capabilities within the application. Below are the details of the implementation and its supported functionalities:

## Simple Usage

```python
from docarray.index import MongoAtlasDocumentIndex
import numpy as np

class MyDoc(BaseDoc):
    text: str
    embedding: NdArray[10]

docs = [MyDoc(text=f'text {i}', embedding=np.random.rand(10)) for i in range(10)]
query = np.random.rand(10)
db = MongoAtlasDocumentIndex[MyDoc](host='localhost')
db.index(docs)
results = index.find(query, search_field='embedding', limit=10)
```

## Supported Functionality
- **Find (vector search):** Enables vector-based search.
- **Filter:** Allows filtering on textual and numeric data using MongoDB syntax.
- **Text Search:** Supports text search using regex match.
- **Get/Del:** Retrieve and delete operations.
- **Subindex:** Ability to create subindexes for better organization.

## Integration Tests and documentation
- `tests/index/mongo_atlas` 
- `docs/API_reference/doc_index/backends/mongodb.md`

## Coming soon
- Implementation of QueryBuilder and Hybrid Search.

### CODE REVIEW

**Comment by JoanFM** [comment]
Hello @caseyclements ,

Thanks a lot for this amazing contribution. May I ask if you could please sign off the commits so that we can pass the DCO and merge the PR if accepted?

File: docarray/index/backends/mongodb_atlas.py [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+
+    @staticmethod
+    def _mongo_to_doc(mongo_doc: dict) -> dict:
+        result = mongo_doc.copy()
+        result["id"] = result.pop("_id")
+        score = result.pop("score", None)
+        return result, score
+
+    @staticmethod
+    def _mongo_to_docs(mongo_docs: Generator[Dict, None, None]) -> List[dict]:
```
Comment (JoanFM) [inline docarray/index/backends/mongodb_atlas.py:200]: the type annotation seems wrong

File: docarray/index/backends/mongodb_atlas.py [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+                result[name] = list(result[name])
+
+        result["_id"] = result.pop("id")
+        return result
+
+    def _docs_to_mongo(self, docs):
+        return [self._doc_to_mongo(doc) for doc in docs]
+
+    @staticmethod
+    def _mongo_to_doc(mongo_doc: dict) -> dict:
```
Comment (JoanFM) [inline docarray/index/backends/mongodb_atlas.py:193]: wrong type annotation

File: docarray/index/backends/mongodb_atlas.py [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+        Establish a connection to MongoDB Atlas.
+        """
+
+        client = MongoClient(
+            atlas_connection_uri,
+            # driver=DriverInfo(name="docarray", version=version("docarray"))
+        )
+        return client
+
+    def _create_indexes(self):
```
Comment (JoanFM) [inline docarray/index/backends/mongodb_atlas.py:88]: this is not implemented?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: docarray/index/backends/mongodb_atlas.py [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+
+    @staticmethod
+    def _mongo_to_doc(mongo_doc: dict) -> tuple[dict, float]:
+        result = mongo_doc.copy()
+        result["id"] = result.pop("_id")
+        score = result.pop("score", None)
+        return result, score
+
+    @staticmethod
+    def _mongo_to_docs(mongo_docs: Generator[Dict, None, None]) -> tuple[list[dict], list[float]]:
```
Comment (JoanFM) [inline docarray/index/backends/mongodb_atlas.py:194]: `tuple` should be changed to `Tuple` from `typing`. This is causing tests to fail

File: docarray/index/backends/mongodb_atlas.py [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+                result[name] = list(result[name])
+
+        result["_id"] = result.pop("id")
+        return result
+
+    def _docs_to_mongo(self, docs):
+        return [self._doc_to_mongo(doc) for doc in docs]
+
+    @staticmethod
+    def _mongo_to_doc(mongo_doc: dict) -> tuple[dict, float]:
```
Comment (JoanFM) [inline docarray/index/backends/mongodb_atlas.py:187]: `tuple` should be changed to `Tuple` from `typing`. This is causing tests to fail

**Review by JoanFM — CHANGES_REQUESTED** [review]
Can you please check how to pass the `black` check? You can find in our `CONTRIBUTING` guidelines the steps to find

File: docarray/index/backends/mongodb_atlas.py (hunk 3 above)
Comment (prakul) [reply docarray/index/backends/mongodb_atlas.py:86]: creating indexes programatically is not available on free tier currently, hence skipping it

File: docarray/index/backends/mongodb_atlas.py (hunk 5 above)
Comment (caseyclements) [reply docarray/index/backends/mongodb_atlas.py:187]: Hi @JoanFM . I'll fix this up in the afternoon.

**Comment by JoanFM** [comment]
We are gettting these errors in the tests caused by the type annotations , maybe my requests were not ggood enough.

```
docarray/index/backends/mongodb_atlas.py:197: in MongoDBAtlasDocumentIndex
    ) -> Tuple[list[dict], list[float]]:
E   TypeError: 'type' object is not subscriptable
```

**Comment by caseyclements** [comment]
Hi @JoanFM. Thank you for your working with me on this. I created a python 3.8 poetry environment for better coverage. black appears to have changed its mind between versions with regards to the ellipsis `...`. Running from 3.8, it moved it to the following line, where the day before, my 3.11 black moved it to the same one! :) I changed it to `pass`. ¯\_(ツ)_/¯ 

```
-    class RuntimeConfig(BaseDocIndex.RuntimeConfig): ...
+    class RuntimeConfig(BaseDocIndex.RuntimeConfig):
+        ...
```
And now 
```
-    class RuntimeConfig(BaseDocIndex.RuntimeConfig): ...
+    class RuntimeConfig(BaseDocIndex.RuntimeConfig):
+        pass
```

I'll turn to the typing issues, `mypy` with py3.8.

**Comment by caseyclements** [comment]
I'm sorry that I missed `list` to `List`. I recognized the required change while updating tuple, but it was a busy morning. I pushed a new commit.

**Comment by caseyclements** [comment]
Hi @JoanFM . I can address the code coverage by adding a new action to `.github/workflows/ci.yml`. @prakul We can set up the correct credentials and share a google doc.

**Comment by JoanFM** [comment]
> Hi @JoanFM . I can address the code coverage by adding a new action to `.github/workflows/ci.yml`. @prakul We can set up the correct credentials and share a google doc.

no need to worry about code coverage

**Comment by JoanFM** [comment]
> > > Hi @JoanFM . I can address the code coverage by adding a new action to `.github/workflows/ci.yml`. @prakul We can set up the correct credentials and share a google doc.
> > 
> > 
> > no need to worry about code coverage
> 
> Cool. What remains then?
> 
> In the next two weeks (I'm in London this coming one) when we add the QueryBuilder, we'll also set up the testing on your end. We are already running on our CI against Atlas on our end. Maybe we could set up a face-to-face meeting in a couple weeks. We can optimize to get the most of MongoDB's API once we know the scope of the use cases for the Indexes, and the data types.

there seems to be a test timing out but not sure if it comes from your changes.

So what is the plan for the upcoming changes feom your side?

**Review by JoanFM — APPROVED** [review]

**Pull request merged**